Dt/oc 1916/pubkey

doubt72 opened the pull request:
Changes for named principal endpoint to allow pushy authentication

Commit:
Roughing out a pubkey lookup API

Commit:
Renamed pubkey endpoint to match named status

Commit:
Stripped unused code from endpoint

Commit:
Renamed pubkey to principal

Commit:
Renamed pubkey to principal in source

Commit:
Added routing for principal endpoint

Commit:
Added routing for principals

Commit:
Extract right object type from request for principals

Commit:
Add proper auth_info handling

Commit:
Allow authorized for all users

Commit:
Fixes for correct call to get pubkey from chef_client

Commit:
Moved check for client to helper function

Commit:
Added support for users for principal endpoint

seth commented on src/chef_wm_named_principal.erl:
```diff
@@ -0,0 +1,130 @@
+allowed_methods(Req, State) ->
+    {['GET'], Req, State}.
+
+fetch_client_or_user(DbContext, OrgName, Name) ->
```
chef_db has a fetch_requestor function, can it be used instead?

seth commented on src/chef_wm_named_principal.erl:
```diff
@@ -0,0 +1,130 @@
+            Client
+    end.
+
+validate_request(_Method, Req, #base_state{chef_db_context = DbContext,
```
I don't understand why we're calling fetch_client_or_user here. Are we actually doing any validation of the request? If not, let's organize it differently.

seth commented on src/chef_wm_named_principal.erl:
```diff
@@ -0,0 +1,130 @@
+      {<<"pubkey">>, PubKey},
+      {<<"pubkey_version">>, Version}]}.
+
+to_json(Req, #base_state{resource_state =
```
You won't need pubkey_version, let's remove it. I was expecting this resource to return a JSON structure that "normalizes" users and clients -- I guess a "principal". Was this discussed and decided against?

seth commented on src/chef_wm.erl:
```diff
@@ -26,6 +26,7 @@
 -type container_name() :: cookbook |
                           data |
                           environment |
+                          principal |
```
This seems wrong in principle :) We aren't going to have a principal container, are we?

Commit:
Changed principal validation to use fetch_requestor instead

Commit:
Changed output to remove pubkey_version and add type

seth commented on include/chef_wm.hrl:
```diff
@@ -180,6 +180,12 @@
           partial_paths = []
          }).
 
+-record(user_state, {
```
Based on our discussion, we should change this to be something like `principal_state`.

Commit:
Added principal_state record

Commit:
Refactored principal endpoint to remove auth and switch to principal record

seth commented on src/chef_wm_named_principal.erl:
```diff
@@ -0,0 +1,113 @@
+                                  ResourceState#principal_state{principal = User}}}
+    end.
+
+validate_request(_Method, Req, State) ->
```
FWIW, I think you don't have to implement this callback if everything is valid.

doubt72 replied:
It violates the chef_wm behavior if you don't (also, it does actually go boom).  Same with auth_info.

seth replied:
Oh, indeed. I confused myself. Was thinking of the wm malformed_request callback. Carry on :)

seth commented on src/chef_wm_named_principal.erl:
```diff
@@ -0,0 +1,113 @@
+    {authorized, Req, State}.
+
+assemble_pubkey_ejson(#chef_user{username=Name, public_key=PubKey}) ->
+    {[{<<"name">>, Name},
```
Since the return values are essentially the same, I would factor out a `principal_type` function and put the ejson literals in one body.

seth commented on src/chef_wm_named_principal.erl:
```diff
@@ -0,0 +1,113 @@
+      {<<"pubkey">>, PubKey},
+      {<<"type">>, <<"client">>}]}.
+
+to_json(Req, #base_state{resource_state =
```
I don't think you need these two fun heads. Code should be the same whether the principal is a client our a user.

Commit:
Unified to_json; code was actually the same for both patterns

Commit:
Refactored principal_state to store info instead of chef objects

Commit:
Removed header signing checks.

There's no point in requiring signed headers if we're not doing any
authentication in the first place; it requires a lot of work on the
pushy side for no gain.
